### DISCUSSION TITLE
useStore syntax question, memoization technique

### DISCUSSION THREAD

**Discussion by bmm-brobbins** [pmndrs/zustand#1386] [Q&A] [answered]
Hi, I've been experimenting with Zustand.  I'm experiencing that using the following syntax results in the component re-rendering upon any change to any of the store's state:

`const [count, setCount] = useStore((state) => [state.count, state.setCount]);`

Whereas splitting apart the getter and setter assignments has the desired result of only re-rendering upon change to the specific selection (state.count):

```
const count = useStore((state) => state.count);
const setCount= useStore((state) => state.setCount);
```

My question: Is there a way to use something like the first syntax and also avoid any unnecessary render cycles?  I've also tried wrapping it in a useCallback, to no avail:

`const [count, setCount] = useStore(useCallback((state) => [state.count, state.setCount],[]));`

No big deal having them split apart, but I'm kind of fond of the combined syntax, as it's a drop-in replacement for useState();  I've also seen this syntax in the docs or examples somewhere.

Thank you.

Btw, I've created the store in its own file, and then importing its useStore into my components.  Maybe I'm doing it wrong?

```
// zustandStore.js

import create from "zustand";
import produce from "immer";

export const useStore = create((set) => (
    {
        count: 0, 
        text: 'hello', 
        myObj: {text: 'myObjText'}, 
        myImmerizedObj: {text: 'myImmerizedObjText'},
        setCount: (newCount) => set(() => ({ count: newCount })),
        setText: (newText) => set(() => ({ text: newText })),
        setMyObjText: (newText) => set((state) => ({ myObj : {...state.myObj, text: newText} })),
        setMyImmerizedObjText: (newText) => set(produce((state) => { state.myImmerizedObj.text = newText })),
    }
));
```
And a sample component (all the useCallback's seem superfluous):
```
import {useStore} from "./zustandStore.js";
import {useCallback, useMemo} from "react";

const Counter = () => {
    // Good. Component only re-renders if change to state.count
    const count = useStore(useCallback((state) => state.count,[]));
    const setCount= useStore(useCallback((state) => state.setCount,[]));

    // Not as good. Component will re-render with a change to any of the store's state:
    // const [count, setCount] = useStore(useCallback((state) => [state.count, state.setCount],[]));

    const inc = () => {
        console.log("Updating state.count");
        setCount(count + 1);
    }

    console.log("Rendering Counter");
    return (
        <div>
            {count} <button onClick={inc}>+1</button>
        </div>
    );
}
```
Let me know and I can post the entire app if that would be helpful to answering the question.

**Comment by dai-shi** [comment, marked as answer]
> My question: Is there a way to use something like the first syntax and also avoid any unnecessary render cycles?

If you prefer the first syntax, you need to specify an equality function.
```js
const [count, setCount] = useStore((state) => [state.count, state.setCount], shallow);
```
Otherwise, it will create a new array every time, which causes extra renders.

**Reply by bmm-brobbins** [reply]
Cool.  And thanks for the quick reply.
